Allow opting out of getting `LocalChain` updates with `FullScanRequest`/`SyncRequest` structures

evanlinjin opened the pull request:
Closes #1528

### Description

Some users use `bdk_esplora` to update `bdk_chain` structures *without* a `LocalChain`. ~~This PR introduces "low-level" methods to `EsploraExt` and `EsploraAsyncExt` which populates a `TxGraph` update with associated data.~~

We change `FullScanRequest`/`SyncRequest` to take in the `chain_tip` parameter as an option. Spk-based chain sources (`bdk_electrum` and `bdk_esplora`) will not fetch a chain-update if `chain_tip` is `None`, allowing callers to opt-out of receiving updates for `LocalChain`.

We change `FullScanRequest`/`SyncRequest` to have better ergonomics when inspecting the progress of syncing (refer to #1528).

We change `FullScanRequest`/`SyncRequest` to be constructed with a builder pattern. This is a better API since we separate request-construction and request-consumption.

Additionally, much of the `bdk_esplora` logic has been made more efficient (less calls to Esplora) by utilizing the `/tx/:txid` endpoint (`get_tx_info` method). This method returns the tx and tx_status in one call. The logic for fetching updates for outpoints has been reworked to support parallelism.

Documentation has also been updated.

### Notes to reviewers

This PR has evolved somewhat. Initially, we were adding more methods on `EsploraExt`/`EsploraAsyncExt` to make syncing/scanning more modular. However, it turns out we can just make the `chain_tip` parameter of the request structures optional. Since we are changing the request structures, we might as well go further and improve the ergonomics of the whole structures (refer to #1528). This is where we are at with this PR.

Unfortunately, the changes are now breaking. I think this is an okay tradeoff for the API improvements (before we get to stable).

### Changelog notice

* Change request structures in `bdk_chain::spk_client` to be constructed via a builder pattern, make providing a `chain_tip` optional, and have better ergonomics for inspecting progress while syncing.
* Change `bdk_esplora` to be more efficient by reducing the number of calls via the `/tx/:txid` endpoint. The logic for fetching outpoint updates has been reworked to support parallelism.
* Change `bdk_esplora` to always add prev-txouts to the `TxGraph` update.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
feat(esplora): make ext traits more flexible

Some users would like to use esplora updates with custom `ChainOracle`s
(not `LocalChain`). We introduce "low-level" extension methods that
populate an update `TxGraph` with associated data.

Additionally, much of the logic has been made more efficient. We make
use of the `/tx/:txid` endpoint (`get_tx_info` method) to do a single
call to get both the tx and tx_status. If the tx already exists, we only
fetch the tx_status. The logic for fetching data based on outpoints has
been reworked to be more efficient and it now supports parallelism
Additionally, the logic for fetching data based on outpoints has been
reworked to be more efficient and it now supports parallelism.

Documentation has also been reworked.

Note that this is NOT a breaking change because the API of `full_scan`
and `sync` are not changed.

storopoli commented on crates/esplora/src/lib.rs:
```diff
@@ -1,19 +1,59 @@
+//! be used directly to determine confirmation statuses of each transaction. This is because a
+//! [`LocalChain`] update is contained in the returned update structures. However, sometimes the
+//! caller wishes to use a custom [`ChainOracle`] implementation (something other than
+//! [`LocalChain`]). The following methods ONLY populates an update [`TxGraph`].
```
```suggestion
//! [`LocalChain`]). The following methods ONLY populates an update [`TxGraph`]:
```

storopoli commented on crates/esplora/src/lib.rs:
```diff
@@ -1,19 +1,59 @@
+//!
+//! [`EsploraExt::full_scan`] is designed to be used when importing or restoring a keychain where
+//! the range of possibly used scripts is not known. In this case it is necessary to scan all
+//! keychain scripts until a number (the `stop gap`) of unused scripts is discovered.
```
```suggestion
//! keychain scripts until a number (the `stop_gap`) of unused scripts is discovered.
```

storopoli reviewed (commented):
Some high-level crate doc comments.
I might go over the code until the end of the weekend.

notmandatory commented:
Since this is a refactor and won't break the `bdk_wallet` API I'm not going to assign it a milestone until we get together for a team backlog prioritization call.

LagginTimes commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -112,6 +164,235 @@ impl EsploraExt for esplora_client::BlockingClient {
+                        }
+                    }
+                    EsploraResp::Tx(Some(tx_info)) => {
+                        let _ = tx_graph.insert_tx(tx_info.to_tx());
```
There was a discussion when the fee calculation for Electrum was implemented, that when we populate by Txids and Outpoints, we still want to add the prev TxOuts for fee calculation because there might be new relevant data. I just noticed that both the previous and new Esplora implementations appear to not actually do this, and was wondering if this needs to be added in?

evanlinjin replied:
Added.

LagginTimes commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -112,6 +164,235 @@ impl EsploraExt for esplora_client::BlockingClient {
+        .map(|_| ())
+    }
+
+    fn populate_using_txids<I: IntoIterator<Item = Txid>>(
```
This is probably just a personal nit, but I'm of the opinion that the codebase would look "neater" if the naming scheme of conceptually similar methods were the same across both Electrum and Esplora, i.e. match `bdk_electrum`'s `populate_with_txids`.

evanlinjin replied:
Thanks for the suggestion. The reason why I have a different name is because we aren't exactly populating the update graph with txids. Instead, we are using txids to fetch txs & anchors for the tx graph. However, I do feel the name can be improved. Let me know if you have any suggestions.

LagginTimes replied:
I think I lean towards something similar to `chain_update`, maybe `graph_update_with_txids` or something similar? What do you think?

LLFourn replied:
it's not populating, it's fetching transactions from a server. The fact that it's doing IO and fertching txs should be emphasized.

evanlinjin replied:
I mean technically the method is doing both? It's fetching txs and tx metadata from Esplora and then populating a `TxGraph` with it?

However, I'm thinking it's better to change the method signatures to **output** a `TxGraph` instead - this creates a nicer API, and the performance sacrifice is not too much. I had `&mut TxGraph` as an input because that was how the internal logic worked and it "felt more efficient" (which it is, but not by much).

Then we can change the methods to `fetch_txs_of_txids`, `fetch_txs_of_outpoints`, `fetch_txs_of_spks`, `fetch_txs_of_keychain_spks`, etc.

evanlinjin replied:
Changed the methods to be as mentioned above. Except I've decided to use `with` as a preposition (instead of `of`).

evanlinjin commented:
> Since this is a refactor and won't break the `bdk_wallet` API I'm not going to assign it a milestone until we get together for a team backlog prioritization call. 

That is fine for me. Because some users are waiting on this, I would appreciate it if we have it in soon. It doesn't break any API since it only adds new exposed methods.

LLFourn commented:
Concept ACK

LLFourn commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -10,59 +10,108 @@ use bdk_chain::{
 use crate::anchor_from_status;
 
-/// [`esplora_client::Error`]
+/// [`esplora_client::Error`] wrapped in a [`Box`].
```
you don't need to say it's wrapped in a `Box`. No one cares about this and it's right there in the type definition.

ValuedMammal replied:
we may not need to box the error anymore after rust-esplora-client moved to using minreq

evanlinjin committed:
feat(esplora): always fetch prevouts

Prevouts are needed to calculate fees for transactions. They are
introduced as floating txouts in the update `TxGraph`. A helper method
`insert_prevouts` is added to insert the floating txouts using the
`Vin`s returned from Esplora.

Also replaced `anchor_from_status` with `insert_anchor_from_status` as
we always insert the anchor into the update `TxGraph` after getting it.

Also removed `bitcoin` dependency as `bdk_chain` already depends on
`bitcoin` (and it's re-exported).

LLFourn commented on crates/chain/src/changeset.rs:
```diff
@@ -1,5 +1,4 @@
 /// A changeset containing [`crate`] structures typically persisted together.
-#[cfg(feature = "miniscript")]
```
If it's not necessary part of the PR revert it.

The correct way to clean this up is to remove it. If we're not going to remove it yet at least call the file `wallet_changeset.rs` and the type `WalletChangeSet` with a big notice at the top saying "This is going to be moved soon". It's confusing to all it `changeset.rs` because people will want to know how to create their own changeset but the trait for that is actually called `Append`.

evanlinjin replied:
Removed the commit for now!

LLFourn commented on crates/esplora/src/lib.rs:
```diff
@@ -1,22 +1,63 @@
+//! the range of possibly used scripts is not known. In this case it is necessary to scan all
+//! keychain scripts until a number (the `stop_gap`) of unused scripts is discovered.
+//!
+//! For a sync or full scan, the user receives relevant blockchain data and output updates for
```
This sentence needs to be reworked. The user receives it? (I think you mean caller) and how exactly do they output updates. Actually I thought the blockchain data was the update. Probably just delete all this with the other suggestion.

evanlinjin replied:
Removed.

LLFourn commented on crates/esplora/src/lib.rs:
```diff
@@ -1,22 +1,63 @@
+//! The two primary methods are [`EsploraExt::sync`] and [`EsploraExt::full_scan`].
 //!
-//! Refer to [`example_esplora`] for a complete example.
+//! [`EsploraExt::sync`] is used to sync against a subset of wallet data. For example, transaction
```
All this documentation seems redundant since we have `SynRequest` and `FullScanRequest`.

> This library is primarily intended to satisfy `SyncRequest`s with [`sync`] and [`FullScanRequest`]s with [`full_scan`]. 

To know what that means they can read the documentation for those types.

evanlinjin replied:
Removed.

LLFourn reviewed (dismissed):
Small doc improvements requested.

LLFourn commented:
Hmm I wonder what would happen if we just made updating the chain optional as part of `SyncRequest/SyncResult`?

evanlinjin committed:
docs(esplora): Simplify crate docs

oleonardolima reviewed (commented):
Concept ACK

ValuedMammal commented on crates/esplora/src/lib.rs:
```diff
@@ -1,22 +1,46 @@
+//! # Async
+//!
+//! Just like how [`EsploraExt`] extends the functionality of an
+//! [`esplora_client::BlockingClient`], [`EsploraExt`] is the async version which extends
```
```suggestion
//! [`esplora_client::BlockingClient`], [`EsploraAsyncExt`] is the async version which extends
```

evanlinjin committed:
feat!: rework `FullScanRequest` and `SyncRequest`

Change `FullScanRequest` and `SyncRequest` take in a `chain_tip` as an
option. In turn, `FullScanResult` and `SyncResult` are also changed to
return the update `chain_tip` as an option. This allows the caller to
opt-out of getting a `LocalChain` update.

Rework `FullScanRequest` and `SyncRequest` to have better ergonomics
when inspecting the progress of items of a sync request. Richer progress
data is provided to the inspect closure.

Introduce `FullScanRequestBuilder` and `SyncRequestBuilder`. Separating
out request-construction and request-consumption in different structs
simplifies the API and method names.

Simplify `EsploraExt` and `EsploraAsyncExt` back to having two methods
(`full_scan` and `sync`). The caller can still opt out of fetching a
`LocalChain` update with the new `FullScanRequest` and `SyncRequest`.

evanlinjin committed:
fix: no premature collect

evanlinjin committed:
docs(esplora): Fix typo

LLFourn commented on crates/esplora/src/async_ext.rs:
```diff
@@ -19,56 +20,102 @@ type Error = Box<esplora_client::Error>;
+    /// script pubkey that contains a non-empty transaction history.
+    ///
+    /// Refer to [crate-level docs](crate) for more.
+    async fn fetch_txs_with_keychain_spks<I: Iterator<Item = Indexed<ScriptBuf>> + Send>(
```
do we actually need this as part of the API now?

evanlinjin replied:
This is just a private helper-function now.

ValuedMammal commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,552 @@
-    #[must_use]
-    pub fn chain_spks(
+    /// Add [`Script`]s coupled with an associated label that will be synced against.
+    pub fn spks_with_labels(
```
I think an example could help for knowing how to include the `SpkLabel`

```rust
/// # Example
/// 
/// Sync revealed script pubkeys obtained from a [`KeychainTxOutIndex`].
///
/// ```rust
/// # use bdk_chain::spk_client::SyncRequest;
/// # use bdk_chain::indexer::keychain_txout::KeychainTxOutIndex;
/// # let indexer = KeychainTxOutIndex::<()>::default();
/// let indexed_spks = indexer.revealed_spks(..);
/// let request = SyncRequest::builder()
///     .spks_with_labels(indexed_spks)
///     .build();
/// ```
```

ValuedMammal commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,552 @@
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let consumed = self.request.spks_consumed;
+        (consumed, Some(consumed))
+    }
```
Is `size_hint` meant to return the number of items consumed or remaining?

evanlinjin replied:
Good point, it is meant to be remaining.

ValuedMammal reviewed (commented):
I haven't reviewed the whole diff but I can say that tests pass locally for me and the electrum/esplora example crates all work nicely.

notmandatory commented:
Concept ACK since it looks like the only changes to the `bdk_wallet` beta API is renaming `start_sync_with_revealed_spks()` to `build_sync_with_revealed_spks()` and `start_full_scan()` to `build_full_scan()`. Though in general I'd like to keep API changes in this milestone as close as we can to none.

The return type also changes but it isn't something a user would be interacting with unless they're developing their own SPK based blockchain client (which I haven't heard of anyone doing).

evanlinjin commented:
> Concept ACK since it looks like the only changes to the `bdk_wallet` beta API is renaming `start_sync_with_revealed_spks()` to `build_sync_with_revealed_spks()` and `start_full_scan()` to `build_full_scan()`. Though in general I'd like to keep API changes in this milestone as close as we can to none.
> 
> The return type also changes but it isn't something a user would be interacting with unless they're developing their own SPK based blockchain client (which I haven't heard of anyone doing).

Looking at your comment now, I don't think my renaming adds much value, I will change it back.

LagginTimes commented:
LGTM so far, will continue reviewing this tomorrow. 👍

notmandatory commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2319,9 +2322,10 @@ impl Wallet {
+    pub fn build_sync_with_revealed_spks(&self) -> SyncRequestBuilder<(KeychainKind, u32)> {
+        SyncRequest::builder()
+            .chain_tip(self.chain.tip())
+            .revealed_spks_from_indexer(&self.indexed_graph.index, ..)
```
❓ I like the new builder pattern for sync and full scan requests, but do we really need to return the builder instead of the final built request (also for full scan) here in Wallet? My understanding is that we'd add new Wallet methods for other useful request options (ie. build_sync_with_revealed_unused_spks(), or something like that).
```suggestion
            .revealed_spks_from_indexer(&self.indexed_graph.index, ..)
            .build()
```

evanlinjin replied:
I think it's better to return the builder, since the caller can further customize it. I've also changed chain sources to take in `impl Into<...Request>` so they can directly take in a builder struct.

LagginTimes commented on crates/esplora/README.md:
```diff
@@ -26,11 +27,17 @@ To use the extension traits:
 use bdk_esplora::EsploraExt;
-// for async
+// // for async
 // use bdk_esplora::EsploraAsyncExt;
```
Can probably do without the additional commenting.

```suggestion
// for async
use bdk_esplora::EsploraAsyncExt;
```

evanlinjin committed:
docs(chain): improve `SyncRequestBuilder::spks_with_labels` docs

evanlinjin committed:
fix(chain): correct `Iterator::size_hint` impl

evanlinjin committed:
docs(esplora): README example, uncomment async import

notmandatory commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,589 @@
+
+/// Builds a [`SyncRequest`].
+#[must_use]
+pub struct SyncRequestBuilder<SpkLabel = ()> {
```
Is `SpkLabel` essentially the same thing as `I` (index?) in `SyncRequest<I = ()>` ? It's confusing to use different identifiers for generic types that represent the same thing. I'm also worried that the "Label" concept might get confused with BIP-329 wallet labels which are unrelated to your change.  One fix is to rename `SpkLabel` to  `SpkIndex` and the related method to `spks_with_indexes`. Then also update `SyncRequest<SpkIndex = ()>`.  Same for `FullScan`.

evanlinjin replied:
For `FullScan`, what should we name the `K` generic parameter? Would `Keychain` worK?

evanlinjin replied:
I think just having `I` for spk index and `K` for keychain should be okay? I renamed the generic type parameters like so.

notmandatory replied:
Yes `I` for spk index and `K` for keychain makes sense. My main concern is that what ever name we use it's used consistently.  Will you also rename `SkpLabel` on this line and below to `I` ?

ValuedMammal reviewed (commented):
After looking over the changes to esplora, I did confirm it's possible to calculate fee for all wallet transactions after doing a full scan

evanlinjin commented:
@ValuedMammal it should also be possible to calculate fee for all wallet transactions after doing sync

notmandatory reviewed (changes requested):
I tested all the examples (rebased on #1549) and confirmed scan, sync. I have one question/request but otherwise this PR looks good to merge. 

I also have one complaint, this PR should have been split up into at least two PRs, 1. to refactor the electrum and esplora scan and sync and make chain optional, and 2. to refactor sync and scan requests with the builder and more progress details. Big PRs like this are hard to properly review and prioritize.

But since I'd like to get this in for a beta.2 this week we don't have time to split and re-review separate PRs for this one. For all PR authors going forward we need to work on making smaller more focused single purpose PRs.

ValuedMammal commented on crates/esplora/src/async_ext.rs:
```diff
@@ -230,176 +237,235 @@ async fn chain_update<A: Anchor>(
+        client,
+        // NOTE: @evanlinjin We need to collect here otherwise we run into lifetime issues, but I
+        // do not know why. If someone can figure this out, that will be amazing.
+        outpoints.iter().map(|op| op.txid).collect::<Vec<_>>(),
```
not sure if this is what you want, but it seems to compile

```suggestion
        outpoints.iter().copied().map(|op| op.txid),
```

evanlinjin replied:
Thank you @ValuedMammal, that seems to work. I'm still quite confused why your solution works and why `outpoints.iter().map(|op| op.txid)` does not. If someone can explain it to me, that will be wonderful.

ValuedMammal replied:
I found this?

https://users.rust-lang.org/t/implementation-of-send-is-not-general-enough-but-cannot-make-it-more-general/115087

evanlinjin committed:
refactor(chain)!: Rename `spks_with_labels` to `spks_with_indices`

and use consistent generic parameter names across `SyncRequest` and
`SyncRequestBuilder`.

notmandatory commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,586 @@
     }
 
-    /// Chain on additional [`Script`]s that will be synced against.
+    /// Add [`Script`]s coupled with an associated label that will be synced against.
```
```suggestion
    /// Add [`Script`]s coupled with an associated index that will be synced against.
```

notmandatory commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,586 @@
+    inner: SyncRequest<SpkLabel>,
+}
+
+impl<SpkLabel> Default for SyncRequestBuilder<SpkLabel> {
```
```suggestion
pub struct SyncRequestBuilder<I = ()> {
    inner: SyncRequest<I>,
}

impl<I> Default for SyncRequestBuilder<I> {
```

evanlinjin replied:
Oh no! Fixing now

notmandatory reviewed (changes requested):
Looks like a few more spots where `SpkLabel` needs to be renamed to `I`.

ValuedMammal commented on crates/chain/src/spk_client.rs:
```diff
@@ -1,388 +1,586 @@
+    /// indexer.insert_descriptor("descriptor_a", descriptor_a)?;
+    /// indexer.insert_descriptor("descriptor_b", descriptor_b)?;
+    ///
+    /// /* Assume that the caller does more mutations to the `indexer` here... */
```
What extra assumptions are needed for this example to make sense? (I would think none)

ValuedMammal reviewed (approved):
ACK 6d77e2e2c3e7c8d49c50bd6b41e7dbc3ddd1d275

rustaceanrob commented:
It is non-intuitive to me to have `SyncProgress` be a collection of `usize` fields. It seems as though you can "consume" the SPK and still have the request fail? I might be interpreting this diff wrong but I feel as though the `SyncProgress` should be derived from a collection of `SyncItem` rather than being a distinct object

i.e. `impl SyncProgress for Vec<SyncItem>`

notmandatory reviewed (approved):
ACK 6d77e2e2c3e7c8d49c50bd6b41e7dbc3ddd1d275

notmandatory commented:
> It is non-intuitive to me to have `SyncProgress` be a collection of `usize` fields. It seems as though you can "consume" the SPK and still have the request fail? I might be interpreting this diff wrong but I feel as though the `SyncProgress` should be derived from a collection of `SyncItem` rather than being a distinct object
> 
> i.e. `impl SyncProgress for Vec<SyncItem>`

I see your point but in practice the `SyncProgress` will be used to just show the percent completed for these items and especially for mobile apps using language bindings transferring smaller and simpler structs is better.
